staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation/constrained_impersonation.go

```diff
@@ -0,0 +1,252 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package impersonation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	authenticationv1 "k8s.io/api/authentication/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	"k8s.io/apiserver/pkg/endpoints/filters"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/server/httplog"
+	"k8s.io/klog/v2"
+)
+
+// TODO add metrics
+
+// WithConstrainedImpersonation implements constrained impersonation as described in https://kep.k8s.io/5284
+// It also includes a complete reimplementation of legacy impersonation for backwards compatibility.
+// At a high level, constrained impersonation uses multiple authorization checks to allow for the granular
+// expression of impersonation access.  For example, a service account may be authorized to impersonate the
+// node that it is associated with but only when listing pods.  See the linked KEP for further details.
+func WithConstrainedImpersonation(handler http.Handler, a authorizer.Authorizer, s runtime.NegotiatedSerializer) http.Handler {
+	return &constrainedImpersonationHandler{
+		handler: handler,
+		tracker: newImpersonationModesTracker(a),
+		s:       s,
+	}
+}
+
+type constrainedImpersonationHandler struct {
+	handler http.Handler
+	tracker *impersonationModesTracker
+	s       runtime.NegotiatedSerializer
+}
+
+func (c *constrainedImpersonationHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+
+	wantedUser, err := processImpersonationHeaders(req.Header)
+	if err != nil {
+		responsewriters.InternalError(w, req, err)
```

Comment:
I know this matches pre-existing behavior, but the only error return looks like it is caused by a bad request ... it's weird to respond with a 500 for that when it was the requester's fault. I'd suggest switching to BadRequest (and likely switching legacy impersonation to respond with BadRequest as well)